### ISSUE TITLE
Move cfreactor into twisted.internet

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7590]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#312|
|Type|enhancement|
|Created|2003-10-04 06:06:10Z|

<details><summary>Searchable metadata</summary>

```
trac-id__312 312
type__enhancement enhancement
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1065247570000000 1065247570000000
changetime__1468236888366830 1468236888366830
version__ 
owner__etrepum etrepum
cc__itamarst cc__etrepum
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
However, you do need to change the code so the interfaces match existing
reactors, i.e. users just do reactor.run(). this could  be done by e.g. having
reactor.run just call the CF event loop  to run and have a callback on that that
does equivalent of current reactor.run.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="etrepum's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|etrepum commented|
|-|-|

```
#!html
<pre>
I can't change the code.  By default it will choose the current CFRunLoop, which may or may not be 
the correct CFRunLoop.  CFRunLoops are stacked; asking for the current runloop will 
give you the bottom of the stack.  This is not always ideal, for example when running a Cocoa 
application you will want the [NSRunLoop currentRunLoop], with Carbon you will want something 
else, and with a pure CoreFoundation application you may or may not want the default run loop.
There must be a way to specify which run loop you would like to integrate with.  I'm open to 
suggestions as for other places to specify which runloop you would like to integrate with.  That 
said, reactor.run() with no arguments will give you a runloop, but perhaps not the correct one.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Argh. Even so, I'd prefer you specify it via install() rather than run().
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Hmm, I should clarify. In all reactors you install and then reactor.run(), and
reactor.stop() to stop. You shouldn't have to run the gui event loop separately.
Twisted code should be as platform neutral as possible, so the GUI integration
bit should be isolated to the reactor installation phase.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="etrepum's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|etrepum commented|
|-|-|

```
#!html
<pre>
The only issue I see with this is that you have to install() before you really do anything with 
Twisted, you may not have a runloop by then.  The way a Cocoa application starts is that it starts 
up (after you give up control with NSApplicationMain(), or whatever) and you get a callback when 
all of your classes ready to run, you'd have to import twisted and install the 
reactor at that point and no earlier.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Double argh. If this is how it has to work then this is how it has to work.

However, we do need (or at least want) to allow tests to run under this, so we
can get a buildslave working. So trial -r cocoa ought to work.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Duplicate of bug 164, so I'm closing it.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|GitHub commented|
|-|-|

In changeset 09d06350a3b5a18d03ec6a15bf6b8857b70b84a7
```
#!CommitTicketReference repository="" revision="09d06350a3b5a18d03ec6a15bf6b8857b70b84a7"
Merge pull request #312 from twisted/8582-rodrigc-iutils-py3

Author: rodrigc
Reviewer: Lukasa
Fixes: #8582

Increase test coverage of test_utils.py on Python 3
```
